File: dotnet/Multiplayer/Games/Games.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Michitai.Multiplayer.Games;

namespace Michitai.Multiplayer.Games
{
    /// <summary>
    /// Provides methods for game-level operations including player listings and global game data management.
    /// </summary>
    public class Games
    {
        /// <summary>
        /// Retrieves a list of all players in the game.
        /// Requires the private API token for authentication.
        /// </summary>
        /// <param name="client">The API client instance.</param>
        /// <param name="ct">Cancellation token for the async operation.</param>
        /// <returns>Response containing the list of all players with their basic information.</returns>
        public static Task<PlayerListResponse> GetAllPlayers(Client client, CancellationToken ct = default)
            => client.Send<PlayerListResponse>(HttpMethod.Get, client.PrivateUrl(Endpoints.GamePlayersList), null, ct);


        /// <summary>
        /// Retrieves global game data with typed deserialization support.
        /// </summary>
        /// <typeparam name="T">The type to deserialize the game data into.</typeparam>
        /// <param name="client">The API client instance.</param>
        /// <param name="ct">Cancellation token for the async operation.</param>
        /// <returns>Response containing the game data deserialized into the specified type.</returns>
        public static Task<GameDataResponse<T>> GetGameData<T>(Client client, CancellationToken ct = default) where T : class, new()
            => client.Send<GameDataResponse<T>>(HttpMethod.Get, client.Url(Endpoints.GameDataGameGet), null, ct);

        /// <summary>
        /// Updates global game data with the provided object.
        /// Requires the private API token for authentication.
        /// </summary>
        /// <typeparam name="T">The type of the game data object.</typeparam>
        /// <param name="client">The API client instance.</param>
        /// <param name="data">The game data object to update.</param>
        /// <param name="ct">Cancellation token for the async operation.</param>
        /// <returns>Success response confirming the update.</returns>
        public static Task<SuccessResponse> UpdateGameData<T>(Client client, T data, CancellationToken ct = default) where T : class, new()
            => client.Send<SuccessResponse>(HttpMethod.Put, client.PrivateUrl(Endpoints.GameDataGameUpdate), data, ct);
    }
}

File: python/multiplayer/games/games.py
"""Game-level operations: player list and global game data."""

from __future__ import annotations

import dataclasses
from typing import Any, Optional, Type, TypeVar

from .._serialization import from_dict
from ..api_response import SuccessResponse
from ..client import Client
from .models import GameDataResponse, PlayerListResponse

T = TypeVar("T")

# Endpoints
GAME_PLAYERS_LIST = "game_players.php/list"
GAME_DATA_GAME_GET = "game_data.php/game/get"
GAME_DATA_GAME_UPDATE = "game_data.php/game/update"


class Games:
    """Game-level operations (mostly admin — require the private token)."""

    def __init__(self, client: Client) -> None:
        self._client = client

    def get_all_players(self) -> PlayerListResponse:
        """List all players in the game. Requires the private API token."""
        return self._client.send(
            "GET",
            self._client.private_url(GAME_PLAYERS_LIST),
            None,
            PlayerListResponse,
        )

    def get_data(self, data_cls: Optional[Type[T]] = None) -> GameDataResponse:
        """Retrieve global game data.

        :param data_cls: Optional dataclass type to deserialize ``data`` into.
        """
        response = self._client.send(
            "GET",
            self._client.url(GAME_DATA_GAME_GET),
            None,
            GameDataResponse,
        )
        if (
            data_cls is not None
            and dataclasses.is_dataclass(data_cls)
            and isinstance(response.data, dict)
        ):
            response.data = from_dict(data_cls, response.data)
        return response

    def update_data(self, data: Any) -> SuccessResponse:
        """Update global game data. Requires the private API token."""
        return self._client.send(
            "PUT",
            self._client.private_url(GAME_DATA_GAME_UPDATE),
            data,
            SuccessResponse,
        )
